Testify-Based Integration Test Suite for DB Tests (#886)
kyma-incubator/reconciler: docs/integration-testing.md

# Integration Testing

Integration testing is the phase in software testing in which individual software modules are combined and tested as a group. 
Integration testing is conducted to evaluate the compliance of a system or component with specified functional requirements. 
It occurs after unit testing and before system testing. Integration testing takes as its input modules that have been unit tested, groups them in larger aggregates, applies tests defined in an integration test plan to those aggregates, and delivers as its output the integrated system ready for system testing.

> To group these integration test aggregates, we make use of test suites. In software development, a test suite, less commonly known as a validation suite, is a collection of test cases that are intended to be used to test a software program to show that it has some specified set of behaviours. 

Author: ```suggestion
To group these integration test aggregates, we use test suites. A test suite is a collection of test cases that show that a software behaves in a specified way.
```

> A test suite often contains detailed instructions or goals for each collection of test cases and information on the system configuration to be used during testing. 

Reviewer: That's what wiki says and rather vague. Do we want to give more specific guidance for Kyma purposes? 

```suggestion
For each test suite, you should provide the following information:
- Detailed instructions and goals
- The system configuration to be used during testing
- Prerequisite states or step
- Information about the following tests 
```

> A group of test cases may also contain prerequisite states or steps, and descriptions of the following tests.

Reviewer: ```suggestion
```
Incorporated in the preceding list.

## Database Integration Testing (Postgres)

> To run our test suite we use the go library [testify](https://pkg.go.dev/github.com/stretchr/testify/suite), specifically the suite capability. 

Reviewer: ```suggestion
To run our test suite, we use the go library [testify](https://pkg.go.dev/github.com/stretchr/testify/suite), specifically the suite capability. 
```

> Compared to regular testing, this enables us to manage lifecycle of tests run inside a suite through a set of predefined methods (for example, run before or after all or individual tests are executed). 

Reviewer: ```suggestion
With this setup, we can manage the lifecycle of tests run inside a suite with a set of predefined methods (for example, run before or after all or individual tests are executed). 
```

> As a result, it enables us to manage dependant components of the reconciler, in this case our database.

Reviewer: ```suggestion
As a result, we can manage dependent components of the reconciler, in this case, our database.
```

> ### How to write a Postgres Integration Test

Reviewer: For your headlines, you're switching between title case and sentence case. 
[Kyma guidelines](https://kyma-project.io/community/guidelines/content-guidelines/03-formatting/#headings) want sentence case, [SAP guidelines](https://help.sap.com/viewer/DRAFT/e33c591ae4494a659a3f5f983c9d1161/PROD/en-US/1c06558f75ef45c686c83e6e73195142.html) want title case. As this is internal documentation, pick what you like as long as it's consistent.

Author: ```suggestion
### How to write a Postgres integration test
```

> To write your own test based on Postgres, there is a helper available that you can use:

Reviewer: ```suggestion
To write your own test based on postgres, use the following helper:
```

Why is Postgres capitalized here?

Author: I figured its gonna be capitalized everytime if you refer to the solution.

Author: ```suggestion
To write your own test based on PostgreSQL, use the following helper:
```

`pkg/db/transaction_aware_database_container_test_suite.go`

> This suite is able to manage a container lifecycle of a postgres database with the help of the [testcontainers](https://golang.testcontainers.org/) framework, enabling us to couple independent database containers with our test suite execution. 

Reviewer: ```suggestion
This suite is able to manage a container lifecycle of a postgres database with the help of the [testcontainers](https://golang.testcontainers.org/) framework. Thus, we can couple independent database containers with our test suite execution.
```

> It provides the ability to isolate tests running together with the database.

Reviewer: ```suggestion
Furthermore, we can isolate tests running together with the database.
```

> To write your own tests, you won't have to set the test suite up yourself, but you can use established helpers depending on your use case.

Reviewer: ```suggestion
To write your own tests, you won't have to set up the test suite yourself, but you can use established helpers depending on your use case.
```
Like the one in line 20, or other helpers?

> #### Shared Container Integration Tests with automatic Rollback

Reviewer: For your headlines, you're switching between title case and sentence case. 
[Kyma guidelines](https://kyma-project.io/community/guidelines/content-guidelines/03-formatting/#headings) want sentence case, [SAP guidelines](https://help.sap.com/viewer/DRAFT/e33c591ae4494a659a3f5f983c9d1161/PROD/en-US/1c06558f75ef45c686c83e6e73195142.html) want title case. As this is internal documentation, pick what you like as long as it's consistent.

Author: ```suggestion
#### Shared container integration tests with automatic rollback
```

> This setup is nice if you just want to run some business logic of the reconciler against a postgres-database and within one transaction context. 

Reviewer: ```suggestion
Use this setup if you just want to run some business logic of the reconciler against a postgres-database and within one transaction context. 
```

> The test suite takes care of using a shared container that is reused between suites with the same isolation needs and rolls back your changes made in the test instead of committing them automatically.

Reviewer: ```suggestion
The test suite takes care of using a shared container that is reused between test suites with the same isolation needs. It rolls back your changes made in the test instead of committing them automatically.
```

The reuse can be achieved by "leasing" a container suite from a globally managed map.

> *IMPORTANT: You must always return the Lease of the container suite so that the container is cleaned up correctly with `ReturnLeasedSharedContainerTestSuite(*testing.T, settings ContainerSettings)`*

Reviewer: ```suggestion
> **IMPORTANT**: You must always return the lease of the container suite so that the container is cleaned up correctly with `ReturnLeasedSharedContainerTestSuite(*testing.T, settings ContainerSettings)`.
```

```go
package random

import (
	"github.com/kyma-incubator/reconciler/pkg/db"
	"github.com/stretchr/testify/suite"
	"testing"
)

type DbTestSuite struct{ *db.ContainerTestSuite }

func TestDbSuite(t *testing.T) {
	t.Parallel()
	suite.Run(t, &DbTestSuite{db.LeaseSharedContainerTestSuite(
		t, db.DefaultSharedContainerSettings, true, false
	)})
	db.ReturnLeasedSharedContainerTestSuite(t, db.DefaultSharedContainerSettings)
}
```

#### Fully Isolated Integration Tests

If you want a dedicated container for your test (for example, because of special requirements or if you want to have committed changes for something like a collision detection), you can use `IsolatedContainerTestSuite(t *testing.T, debug bool, settings ContainerSettings)`.

> Do not choose this method if you just need simple business logic tests, because this consumes many resources for spinning up a dedicated container.

Reviewer: ```suggestion
Do not choose this method if you just need simple business logic tests, because it consumes many resources for spinning up a dedicated container.
```

# Integration Testing

Integration testing is the phase in software testing in which individual software modules are combined and tested as a
group. Integration testing is conducted to evaluate the compliance of a system or component with specified functional
requirements. It occurs after unit testing and before system testing. Integration testing takes as its input modules
that have been unit tested, groups them in larger aggregates, applies tests defined in an integration test plan to those
aggregates, and delivers as its output the integrated system ready for system testing.

> To group these, we make use of test suites. In software development, a test suite, less commonly known as a validation

Reviewer: "these" what?

suite, is a collection of test cases that are intended to be used to test a software program to show that it has some
specified set of behaviours. A test suite often contains detailed instructions or goals for each collection of test
cases and information on the system configuration to be used during testing. A group of test cases may also contain
prerequisite states or steps, and descriptions of the following tests.

## Database Integration Testing (Postgres)

To run our test suite we use the go library [testify](https://pkg.go.dev/github.com/stretchr/testify/suite),

> specifically the suite capability. Compared to regular testing, this allows us to manage lifecycle of tests run inside a

Reviewer: ```suggestion
specifically the suite capability. Compared to regular testing, this enables us to manage lifecycle of tests run inside a
```

it wasn't forbidden

> suite through a set of predefined methods (run e.g. before or after all or individual tests are executed). This allows

Reviewer: ```suggestion
suite through a set of predefined methods (for example, run before or after all or individual tests are executed). This enables
```

us to manage dependant components of the reconciler, in this case our database.

### How to write a Postgres Integration Test

To write your own test based on Postgres, there is a helper available that you can use:
`pkg/db/transaction_aware_database_container_test_suite.go`

This suite is able to manage a container lifecycle of a postgres database with the help of the

> [testcontainers](https://golang.testcontainers.org/) framework, allowing us to couple independant database containers

Reviewer: ```suggestion
[testcontainers](https://golang.testcontainers.org/) framework, enabling us to couple independent database containers
```

with our test suite execution. This provides the ability to isolate tests running together with the database.

> To write your own Tests, you won't have to set the TestSuite up yourself, but you can use established Helpers depending

Reviewer: why are "Tests" and "TestSuite" and "Helpers" capitalized here? They were not capitalized in the sentences before. Are they special kinds of tests, test suites, helpers, different from the ones mentioned above?

```suggestion
To write your own tests, you won't have to set the test suite up yourself, but you can use established helpers depending
```

on your use case.

#### Shared Container Integration Tests with automatic Rollback

This setup is nice if you just want to run some business logic of the reconciler against a postgres-database and within

> one transaction context. The test suite will take care of using a shared container that is reused between suites with

Reviewer: ```suggestion
one transaction context. The test suite takes care of using a shared container that is reused between suites with
```

> the same isolation needs and will rollback your changes made in the test instead of committing them automatically.

Reviewer: ```suggestion
the same isolation needs and rolls back your changes made in the test instead of committing them automatically.
```

The reuse can be achieved by "leasing" a container suite from a globally managed map.

> *IMPORTANT: Make sure to always return the Lease of the container suite to make sure the container gets cleaned up

Reviewer: ```suggestion
> **IMPORTANT**: You must always return the Lease of the container suite so that the container is cleaned up
```

> correctly with `ReturnLeasedSharedContainerTestSuite(*testing.T, settings ContainerSettings)`*

Reviewer: ```suggestion
correctly with `ReturnLeasedSharedContainerTestSuite(*testing.T, settings ContainerSettings)`
```

Author: not sure why you removed that, its starting a special case on line 42. I left it in but made it to one line as asked

```go
package random

import (
	"github.com/kyma-incubator/reconciler/pkg/db"
	"github.com/stretchr/testify/suite"
	"testing"
)

type DbTestSuite struct{ *db.ContainerTestSuite }

func TestDbSuite(t *testing.T) {
	t.Parallel()
	suite.Run(t, &DbTestSuite{db.LeaseSharedContainerTestSuite(
		t, db.DefaultSharedContainerSettings, true, false
	)})
	db.ReturnLeasedSharedContainerTestSuite(t, db.DefaultSharedContainerSettings)
}
```

#### Fully Isolated Integration Tests

> In case you want a dedicated container for your Test (e.g. through special requirements or if you want to have committed

Reviewer: ```suggestion
If you want a dedicated container for your test (for example, because of special requirements or if you want to have committed
```
not sure about the "through".

> changes for something like a collision detection), you can make use of

Reviewer: ```suggestion
changes for something like a collision detection), you can use 
```

> `IsolatedContainerTestSuite(t *testing.T, debug bool, settings ContainerSettings)` to achieve your goal. Do not choose

Reviewer: ```suggestion
`IsolatedContainerTestSuite(t *testing.T, debug bool, settings ContainerSettings)`. Do not choose
```

> this method in case you just need simple business logic tests, since this consumes many resources for spinning up a

Reviewer: ```suggestion
this method if you just need simple business logic tests, because this consumes many resources for spinning up a
```
